sys/linux/init_images.go

```diff
@@ -0,0 +1,240 @@
+	// for two reasons:
+	// 1) It further complicates the already complicated executor code.
+	// 2) We'd need to duplicate the logic in Go for raw image extraction.
+	// So now we do all the initialization in Go and let the C code only interpret the commands.
+	ret, err := parseSyzMountImage(c)
+	if err != nil {
+		deactivateSyzMountImage(c)
+		return
+	}
+	fixUpImageSegments(ret)
+}
+
+type mountImageArgs struct {
+	size          *prog.ConstArg
+	segmentsCount *prog.ConstArg
+	segmentsGroup *prog.GroupArg
+	segments      []*mountImageSegment
+}
+
+func (m *mountImageArgs) filterSegments(filter func(int, *mountImageSegment) bool) {
+	newArgs := []prog.Arg{}
+	newSegments := []*mountImageSegment{}
+	for i, segment := range m.segments {
+		if filter(i, segment) {
+			newSegments = append(newSegments, segment)
+			newArgs = append(newArgs, m.segmentsGroup.Inner[i])
+		}
+	}
+	m.segments = newSegments
+	m.segmentsGroup.Inner = newArgs
+	m.segmentsCount.Val = uint64(len(newArgs))
+}
+
+// Methods for segment sorting.
+func (m *mountImageArgs) Len() int { return len(m.segments) }
+func (m *mountImageArgs) Swap(i, j int) {
+	inner := m.segmentsGroup.Inner
+	inner[i], inner[j] = inner[j], inner[i]
+	m.segments[i], m.segments[j] = m.segments[j], m.segments[i]
+}
+func (m *mountImageArgs) Less(i, j int) bool {
+	if m.segments[i].offset.Val != m.segments[j].offset.Val {
+		return m.segments[i].offset.Val < m.segments[j].offset.Val
+	}
+	return m.segments[i].size.Val < m.segments[j].size.Val
+}
+
+type mountImageSegment struct {
+	data       *prog.DataArg
+	offset     *prog.ConstArg
+	size       *prog.ConstArg
+	parseError error
+}
+
+func parseImageSegment(segmentArg prog.Arg) *mountImageSegment {
+	ret := &mountImageSegment{}
+	segmentFields, ok := segmentArg.(*prog.GroupArg)
+	if segmentFields == nil || !ok {
+		return &mountImageSegment{parseError: fmt.Errorf("it is not a group")}
+	}
+	if len(segmentFields.Inner) != 3 {
+		return &mountImageSegment{parseError: fmt.Errorf("invalid number of nested fields")}
+	}
+	dataPtr, ok := segmentFields.Inner[0].(*prog.PointerArg)
+	if dataPtr == nil || dataPtr.Res == nil || !ok {
+		return &mountImageSegment{parseError: fmt.Errorf("invalid data field ptr")}
+	}
+	ret.data, ok = dataPtr.Res.(*prog.DataArg)
+	if ret.data == nil || !ok {
+		return &mountImageSegment{parseError: fmt.Errorf("invalid data arg")}
+	}
+	ret.size, ok = segmentFields.Inner[1].(*prog.ConstArg)
+	if ret.size == nil || !ok {
+		return &mountImageSegment{parseError: fmt.Errorf("invalid size arg")}
+	}
+	ret.offset, ok = segmentFields.Inner[2].(*prog.ConstArg)
+	if ret.offset == nil || !ok {
+		return &mountImageSegment{parseError: fmt.Errorf("invalid offset arg")}
+	}
+	return ret
+}
+
+func deactivateSyzMountImage(c *prog.Call) {
+	groupArg := c.Args[4]
+	newArg := groupArg.Type().DefaultArg(groupArg.Dir())
+	prog.RemoveArg(groupArg)
+	c.Args[4] = newArg
```

Comment:
I think we also need to set number of segments to 0 so that executor does not crash.